File: queries/h01.R
tpc_h01 <- function(input_func, collect_func = dplyr::collect) {
    res <- input_func("lineitem") %>%
    select(L_SHIPDATE, L_RETURNFLAG, L_LINESTATUS, L_QUANTITY,
           L_EXTENDEDPRICE, L_DISCOUNT, L_TAX) %>%
    # kludge, should be: filter(l_shipdate <= "1998-12-01" - interval x day) %>%
    # where x is between 60 and 120, 90 is the only one that will validate.
    filter(L_SHIPDATE <= as.Date("1998-09-02")) %>%
    select(L_RETURNFLAG, L_LINESTATUS, L_QUANTITY, L_EXTENDEDPRICE, L_DISCOUNT, L_TAX) %>%
    group_by(L_RETURNFLAG, L_LINESTATUS) %>%
    summarise(
      sum_qty = sum(L_QUANTITY),
      sum_base_price = sum(L_EXTENDEDPRICE),
      sum_disc_price = sum(L_EXTENDEDPRICE * (1 - L_DISCOUNT)),
      sum_charge = sum(L_EXTENDEDPRICE * (1 - L_DISCOUNT) * (1 + L_TAX)),
      avg_qty = mean(L_QUANTITY),
      avg_price = mean(L_EXTENDEDPRICE),
      avg_disc = mean(L_DISCOUNT),
      count_order = n()
    ) %>%
    ungroup() %>%
    arrange(L_RETURNFLAG, L_LINESTATUS) %>%
    collect_func()
    return(res)
}

File: ibis_tpc/h01.py
"Pricing Summary Report Query (Q1)"
from .utils import add_date


def tpc_h01(con, DELTA=-90, DATE="1998-12-01"):
    """
    The Pricing Summary Report Query provides a summary pricing report for all
    lineitems shipped as of a given date.  The  date is  within  60  - 120 days
    of  the  greatest  ship  date  contained  in  the database.  The query
    lists totals  for extended  price,  discounted  extended price, discounted
    extended price  plus  tax,  average  quantity, average extended price,  and
    average discount.  These  aggregates  are grouped  by RETURNFLAG  and
    LINESTATUS, and  listed  in ascending  order of RETURNFLAG and  LINESTATUS.
    A  count  of the  number  of  lineitems in each  group  is included.
    """

    lineitem = con.table("lineitem")

    tables = (lineitem,)

    return _tpc_h01(tables, DELTA, DATE)


def _tpc_h01(tables, DELTA=-90, DATE="1998-12-01"):

    lineitem, = tables

    interval = add_date(DATE, dd=DELTA)
    q = lineitem.filter(lineitem.l_shipdate < interval)
    discount_price = lineitem.l_extendedprice * (1 - lineitem.l_discount)
    charge = discount_price * (1 + lineitem.l_tax)
    q = q.group_by(["l_returnflag", "l_linestatus"])
    q = q.order_by(["l_returnflag", "l_linestatus"])
    q = q.aggregate(
        sum_qty=lineitem.l_quantity.sum(),
        sum_base_price=lineitem.l_extendedprice.sum(),
        sum_disc_price=discount_price.sum(),
        sum_charge=charge.sum(),
        avg_qty=lineitem.l_quantity.mean(),
        avg_price=lineitem.l_extendedprice.mean(),
        avg_disc=lineitem.l_discount.mean(),
        count_order=lineitem.count(),
    )
    return q
